### PR TITLE
compile snax alu kernel from linalg

### PR DESCRIPTION
**Pull request by jorendumoulin**
This PR changes the snax-alu kernel by implementing the accelerator in the `accfg` dialect and programming the accelerator within mlir. The codegen is still super hard-coded to one specific configuration, more automation will follow soon.

Nonetheless, I enjoyed the `accfg` experience! Nice work on this

### CODE REVIEW

File: compiler/accelerators/snax.py [hunk 1]
```diff
@@ -17,39 +17,40 @@ class SNAXAccelerator(Accelerator, ABC):
...
-        # Get the launch address, for SNAX should be only one
-        launch_address = None
-        for field, val in acc_op.launch_field_items():
-            assert field == "launch"
-            launch_address = val
-        assert launch_address is not None
-        # Get the launch value,
-        # For SNAX there should only be one value here
-        launch_value = None
+    def lower_acc_launch(launch_op: accfg.LaunchOp, acc_op: accfg.AcceleratorOp) -> Sequence[Operation]:
```
Comment (jorendumoulin) [inline compiler/accelerators/snax.py:20]: I changed this function to be able to handle multiple launch ops

File: kernels/streamer_alu/main.c [hunk 2]
```diff
@@ -1,116 +1,65 @@
...
-    // The DATA_LEN is found in data.h
-    size_t vector_size = DATA_LEN * sizeof(uint64_t);
-    snrt_dma_start_1d(local_a, A, vector_size);
-    snrt_dma_start_1d(local_b, B, vector_size);
-
-    // Measure the end of the transfer process
-    uint32_t end_dma_load = snrt_mcycle();
-  }
+  // Create memref objects for data stored in L3
+  OneDMemrefI32_t memrefA;
```
Comment (jorendumoulin) [inline kernels/streamer_alu/main.c:13]: using I32 here for I64 and I128, as it does not really matter, I think we can change this to OneDMemref_t maybe to simplify things?

File: kernels/streamer_alu/streamer_add.mlir [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#simple_mult_attributes = {
+  indexing_maps = [
+    affine_map<(n) -> (n)>,
+    affine_map<(n) -> (n)>,
+    affine_map<(n) -> (n)>
+  ],
+  iterator_types = ["parallel"],
+  library_call = "snax_alu"
```
Comment (jorendumoulin) [inline kernels/streamer_alu/streamer_add.mlir:8]: pre-dispatched for now

File: kernels/streamer_alu/main.c (hunk 2 above)
Comment (jorendumoulin) [reply kernels/streamer_alu/main.c:14]: In a separate pr, as this affects all of the kernels

Comment (JosseVanDelm) [reply kernels/streamer_alu/main.c:14]: Sounds like a good idea. We could do this with void pointers. Essentially telling the compiler that we are in _expert mode_ :^)

File: kernels/streamer_alu/streamer_add.mlir (hunk 3 above)
Comment (JosseVanDelm) [reply kernels/streamer_alu/streamer_add.mlir:8]: Okay!

**Review by JosseVanDelm — APPROVED** [review]
This looks good @jorendumoulin,
Can you maybe immediately release a new snax version with the 64 bit output and adapt to use that version here first?
Thanks!!!

**Comment by JosseVanDelm** [comment]
> Nonetheless, I enjoyed the accfg experience! Nice work on this

:heart_hands:

File: compiler/accelerators/snax_alu.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        c32 = arith.Constant.from_int_and_width(32, 32)
+        c64 = arith.Constant.from_int_and_width(64, 32)
+
+        ptr_a = memref.ExtractAlignedPointerAsIndexOp.get(op.inputs[0])
+        ptr_b = memref.ExtractAlignedPointerAsIndexOp.get(op.inputs[1])
+        ptr_o = memref.ExtractAlignedPointerAsIndexOp.get(op.outputs[0])
+
+        ptr_a_i32 = builtin.UnrealizedConversionCastOp.get([ptr_a], [builtin.i32])
+        ptr_b_i32 = builtin.UnrealizedConversionCastOp.get([ptr_b], [builtin.i32])
+        ptr_o_i32 = builtin.UnrealizedConversionCastOp.get([ptr_o], [builtin.i32])
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply compiler/accelerators/snax_alu.py:68]: yes haha 🪄

**Pull request merged**